requestIdleCallback can continuously schedule WindowEventLoop's timer

rniwa opened the pull request:
#### cb3478b28c691980974f822d1e662be89a504cf0
<pre>
requestIdleCallback can continuously schedule WindowEventLoop&apos;s timer
<a href="https://bugs.webkit.org/show_bug.cgi?id=282133">https://bugs.webkit.org/show_bug.cgi?id=282133</a>

Reviewed by Wenson Hsieh.

The bug was most likely caused by WindowEventLoop::didReachTimeToRun scheduling itself when exiting early
for shouldEndIdlePeriod returning true. While I have not been able to reproduce the issue, this PR
refactors the code in such a way that a bug like this could never happen.

Prior to this PR, WindowEventLoop::didReachTimeToRun() was responsible for re-scheduling itself. The idea
was that WindowEventLoop will utilize the WindowEventLoop&apos;s main timer to drive the idle callbacks when
there is an idle period. In practice, this led to a fragile design where shouldEndIdlePeriod returning
true can end up in a busy loop between WindowEventLoop::didReachTimeToRun() and Timer.

This PR refactors this code so that there is a dedicated m_idleTimer timer for scheduling idle callback,
and WindowEventLoop::opportunisticallyRunIdleCallbacks and scheduleIdlePeriod are responsible for
scheduling this timer. opportunisticallyRunIdleCallbacks specifically schedules the timer at 1ms after
the next rendering update or next timer which ever is sooner if either is scheduled. Neither is scheduled,
it would schedule immediately trigger a 0s timer.

This PR also removes the knowledge of rendering update time in WindowEventLoop. Instead, WindowEventLoop
will dynamically consults each Document&apos;s Page object for the next rendering update time. It also removes
the flag in OpportunisticTaskScheduler indicating there is an idle callback or not. Instead,
OpportunisticTaskScheduler will dynamically calls document&apos;s hasPendingIdleCallback.

* LayoutTests/requestidlecallback/requestidlecallback-deadline-cap-by-rendering-update.html:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasPendingIdleCallback const): Added.

* Source/WebCore/dom/Document.h:
(WebCore::Document::idleCallbackController const):

* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback): Always schedule idle period when requestIdleCallback
is called.

* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::WindowEventLoop):
(WebCore::WindowEventLoop::scheduleIdlePeriod): Schedule m_idleTimer in 0ms. This is okay because
WindowEventLoop::opportunisticallyRunIdleCallbacks exits early when there are other (higher priority)
tasks and microtasks scheduled in the event loop.
(WebCore::WindowEventLoop::didScheduleRenderingUpdate): Deleted.
(WebCore::WindowEventLoop::didStartRenderingUpdate): Deleted.
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks):
(WebCore::WindowEventLoop::shouldEndIdlePeriod): Now returns true when there are (higher priority) tasks
and microtasks scheduled in the event loop.
(WebCore::WindowEventLoop::computeIdleDeadline):
(WebCore::WindowEventLoop::nextScheduledWorkTime const): Extracted from
(WebCore::WindowEventLoop::nextRenderingTime const): Now polls rendering update time from each Page.
(WebCore::WindowEventLoop::didReachTimeToRun): Removed the much of logic for idle callback.
(WebCore::WindowEventLoop::didFireIdleTimer): Added.

* Source/WebCore/dom/WindowEventLoop.h:

* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::rescheduleIfNeeded):
(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):

* Source/WebCore/page/OpportunisticTaskScheduler.h:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal):
(WebCore::Page::nextRenderingUpdateTimestamp const): Added.
(WebCore::Page::updateRendering):
(WebCore::Page::findMatchingLocalDocument const): Added.
(WebCore::Page::opportunisticallyRunIdleCallbacks):

* Source/WebCore/page/Page.h:
(WebCore::Page::nextRenderingUpdateTimestamp const):

Canonical link: <a href="https://commits.webkit.org/285933@main">https://commits.webkit.org/285933@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b89abac0f8d107a78391ebb09fab27cdb743349

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53657 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27039 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78602 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25466 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76345 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62790 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1442 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58348 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16684 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77295 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48509 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63849 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38758 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45444 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21345 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23799 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66896 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80122 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1545 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/873 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66647 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1690 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63866 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/65922 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9863 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8015 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11464 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1509 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/4297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1538 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1526 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1545 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8462,6 +8462,13 @@ void Document::cancelIdleCallback(int id)
+{
+    if (activeDOMObjectsAreSuspended() || activeDOMObjectsAreStopped())
+        return false;
+    return m_idleCallbackController ? !m_idleCallbackController->isEmpty() : false;
```
We usually like to write these as:
```
return m_idleCallbackController && !m_idleCallbackController->isEmpty();
```

rniwa replied:
Fixed as pointed out by @whsieh.

whsieh commented on LayoutTests/requestidlecallback/requestidlecallback-deadline-cap-by-rendering-update.html:
```diff
@@ -25,7 +25,7 @@
         shouldBeTrue('didRunIdleCallback');
         finishJSTest();
-    }, 200);
+    }, 1000);
```
Would it be reasonable to do something like this, instead of a hard-coded timeout?

```shouldBecomeEqual("didRunIdleCallback", "true", finishJSTest);```

rniwa replied:
Good idea. Will do.

whsieh commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8462,6 +8462,13 @@ void Document::cancelIdleCallback(int id)
+{
+    if (activeDOMObjectsAreSuspended() || activeDOMObjectsAreStopped())
+        return false;
+    return m_idleCallbackController ? !m_idleCallbackController->isEmpty() : false;
```
Nit - _maybe_ reads slightly cleaner this way?

```suggestion
    return m_idleCallbackController && !m_idleCallbackController->isEmpty();
```

rniwa replied:
Sure.

whsieh commented on Source/WebCore/page/Page.cpp:
```diff
@@ -4110,6 +4111,21 @@ void Page::forEachDocument(const Function<void(Document&)>& functor) const
+bool Page::findMatchingLocalDocument(const Function<bool(Document&)>& functor) const
+{
+    for (const auto* frame = &mainFrame(); frame; frame = frame->tree().traverseNext()) {
+        auto* localFrame = dynamicDowncast<LocalFrame>(frame);
```
Should these use `RefPtr`?
```suggestion
    for (RefPtr frame = &mainFrame(); frame; frame = frame->tree().traverseNext()) {
        RefPtr localFrame = dynamicDowncast<LocalFrame>(frame);
```

rniwa replied:
Oh yea, we should probably do that. Fixed.

rniwa committed:
requestIdleCallback can continuously schedule WindowEventLoop's timer
https://bugs.webkit.org/show_bug.cgi?id=282133

Reviewed by Wenson Hsieh.

The bug was most likely caused by WindowEventLoop::didReachTimeToRun scheduling itself when exiting early
for shouldEndIdlePeriod returning true. While I have not been able to reproduce the issue, this PR
refactors the code in such a way that a bug like this could never happen.

Prior to this PR, WindowEventLoop::didReachTimeToRun() was responsible for re-scheduling itself. The idea
was that WindowEventLoop will utilize the WindowEventLoop's main timer to drive the idle callbacks when
there is an idle period. In practice, this led to a fragile design where shouldEndIdlePeriod returning
true can end up in a busy loop between WindowEventLoop::didReachTimeToRun() and Timer.

This PR refactors this code so that there is a dedicated m_idleTimer timer for scheduling idle callback,
and WindowEventLoop::opportunisticallyRunIdleCallbacks and scheduleIdlePeriod are responsible for
scheduling this timer. opportunisticallyRunIdleCallbacks specifically schedules the timer at 1ms after
the next rendering update or next timer which ever is sooner if either is scheduled. Neither is scheduled,
it would schedule immediately trigger a 0s timer.

This PR also removes the knowledge of rendering update time in WindowEventLoop. Instead, WindowEventLoop
will dynamically consults each Document's Page object for the next rendering update time. It also removes
the flag in OpportunisticTaskScheduler indicating there is an idle callback or not. Instead,
OpportunisticTaskScheduler will dynamically calls document's hasPendingIdleCallback.

* LayoutTests/requestidlecallback/requestidlecallback-deadline-cap-by-rendering-update.html:

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::hasPendingIdleCallback const): Added.

* Source/WebCore/dom/Document.h:
(WebCore::Document::idleCallbackController const):

* Source/WebCore/dom/IdleCallbackController.cpp:
(WebCore::IdleCallbackController::queueIdleCallback): Always schedule idle period when requestIdleCallback
is called.

* Source/WebCore/dom/WindowEventLoop.cpp:
(WebCore::WindowEventLoop::WindowEventLoop):
(WebCore::WindowEventLoop::scheduleIdlePeriod): Schedule m_idleTimer in 0ms. This is okay because
WindowEventLoop::opportunisticallyRunIdleCallbacks exits early when there are other (higher priority)
tasks and microtasks scheduled in the event loop.
(WebCore::WindowEventLoop::didScheduleRenderingUpdate): Deleted.
(WebCore::WindowEventLoop::didStartRenderingUpdate): Deleted.
(WebCore::WindowEventLoop::opportunisticallyRunIdleCallbacks):
(WebCore::WindowEventLoop::shouldEndIdlePeriod): Now returns true when there are (higher priority) tasks
and microtasks scheduled in the event loop.
(WebCore::WindowEventLoop::computeIdleDeadline):
(WebCore::WindowEventLoop::nextScheduledWorkTime const): Extracted from
(WebCore::WindowEventLoop::nextRenderingTime const): Now polls rendering update time from each Page.
(WebCore::WindowEventLoop::didReachTimeToRun): Removed the much of logic for idle callback.
(WebCore::WindowEventLoop::didFireIdleTimer): Added.

* Source/WebCore/dom/WindowEventLoop.h:

* Source/WebCore/page/OpportunisticTaskScheduler.cpp:
(WebCore::OpportunisticTaskScheduler::rescheduleIfNeeded):
(WebCore::OpportunisticTaskScheduler::runLoopObserverFired):

* Source/WebCore/page/OpportunisticTaskScheduler.h:

* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdateInternal):
(WebCore::Page::nextRenderingUpdateTimestamp const): Added.
(WebCore::Page::updateRendering):
(WebCore::Page::findMatchingLocalDocument const): Added.
(WebCore::Page::opportunisticallyRunIdleCallbacks):

* Source/WebCore/page/Page.h:
(WebCore::Page::nextRenderingUpdateTimestamp const):

Canonical link: https://commits.webkit.org/285933@main
